### PR TITLE
Add csm-module status

### PR DESCRIPTION
**Pull request by abhi16394**
# Description
Issue - CSM Operator was not calculating status correctly when driver was installed indepedently
It would show as failed even though pods were up and running
solution - the CSM status was not factoring the driver pods to calculate status as it would skip over a condition.
Fixed it by applying a new logic which checks driver status and module status as well
But the code has been changed to calculate the status of all supported modules in dreadnaught.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| https://github.com/dell/csm/issues/1130|

# Checklist:

- [ ] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have maintained backward compatibility

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

1) when driver is installed with no modules installed, since patch release has only powerflex v2.9.0 supported 
   I have tested everything with powerflex-v2.9.0 and i would verify the same with powerflex-v2.9.1 when raising a PR to merge with main branch
   
image

2) when driver is installed with observability and app-mobility installed
image

3) when all the modules are installed successfully but driver is failing 
image

4) when driver is installed successfully but the application-mobility modules are failing 
image

### CODE REVIEW

File: pkg/utils/status.go [hunk 1]
```diff
@@ -654,77 +627,229 @@ func WaitForNginxController(ctx context.Context, instance csmv1.ContainerStorage
...
+				certManagerWebhookRunning = checkFn(&deployment)
+			}
+		}
+	}
+
+	if certEnabled && otelEnabled && metricsEnabled && topologyEnabled {
+		return certManagerRunning && certManagerCainInjectorRunning && certManagerWebhookRunning && otelRunning && metricsRunning && topologyRunning, nil
+	}
+
+	if !certEnabled && otelEnabled && metricsEnabled && topologyEnabled {
```
Comment (jooseppi-luna) [inline pkg/utils/status.go:850]: what if customer has metrics enabled but not otel and topology?

**Review by jooseppi-luna — APPROVED** [review]

**Review by jooseppi-luna — CHANGES_REQUESTED** [review]
meant to request changes not approve

**Comment by alikdell** [comment]
We do not have any unit test code for those go files being changed in this PR. Why we do not have unit test code? I am not even talking about 90 test coverage, basic unit test?

**Comment by abhi16394** [comment]
> We do not have any unit test code for those go files being changed in this PR. Why we do not have unit test code? I am not even talking about 90 test coverage, basic unit test?

Hey @alikdell all the functions inside util folder are called by controller files every time they reconcile. 
so all the tests pertaining to these utils folder files are in csm_controller_test.go file

**Comment by jooseppi-luna** [comment]
@alikdell these functions are currently being tested through csm_controller tests -- I agree that it could be a good idea to have a test suite specifically for the utils package, but as it stands the unit tests in csm_controller do give us sufficient coverage of the utils package -- for us to add tests to utils is much more a nice-to-have, and if we really want it, it should be its own feature. 

This defect could have been avoided if we had run even the most basic E2E tests before release. That's the real test gap that caused this issue, IMO. We need to improve how the E2E tests are run/supported and what is required before merging a PR/posting a release. I'm planning on discussing this in the Monday operator sync-up.

File: pkg/utils/status.go (hunk 1 above)
Comment (jooseppi-luna) [reply pkg/utils/status.go:850]: perfect thanks

**Review by jooseppi-luna — APPROVED** [review]
lgtm

File: pkg/utils/status.go [hunk 2]
```diff
@@ -654,77 +627,245 @@ func WaitForNginxController(ctx context.Context, instance csmv1.ContainerStorage
...
+	veleroEnabled := false
+	certEnabled := false
+	var certManagerRunning bool
+	var certManagerCainInjectorRunning bool
+	var certManagerWebhookRunning bool
+	appMobRunning := false
+	veleroRunning := false
+	var daemonRunning bool
+	readyPods := 0
+	expected := 2
```
Comment (JacobGros) [inline pkg/utils/status.go:642]: Have we tried this on a k8s cluster with worker nodes not equal to 2?
I don't think we should hard code 2 as the expected number here

**Comment by alikdell** [comment]
> @alikdell these functions are currently being tested through csm_controller tests -- I agree that it could be a good idea to have a test suite specifically for the utils package, but as it stands the unit tests in csm_controller do give us sufficient coverage of the utils package -- for us to add tests to utils is much more a nice-to-have, and if we really want it, it should be its own feature.
> 
> This defect could have been avoided if we had run even the most basic E2E tests before release. That's the real test gap that caused this issue, IMO. We need to improve how the E2E tests are run/supported and what is required before merging a PR/posting a release. I'm planning on discussing this in the Monday operator sync-up.

Unit test and E2E are different in my opinion.

**Review by JacobGros — CHANGES_REQUESTED** [review]
I think we should calculate the number of expected pods in daemonset rather than assume it's 2.

File: pkg/utils/status.go (hunk 2 above)
Comment (abhi16394) [reply pkg/utils/status.go]: Thanks for the comments 
i have fixed it and verified it

**Review by JacobGros — APPROVED** [review]

**Pull request merged**